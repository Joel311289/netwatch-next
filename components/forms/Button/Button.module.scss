/* stylelint-disable declaration-block-no-shorthand-property-overrides */
.buttonWrapper {
  @include font($size: 16px, $color: var(--color-dark), $weight: 500);
  @include flex($align: center, $justify: center, $gap: 10px);
  @include box(fit-content, 44px);
  @include no-outline(true);

  font-family: var(--font-text);
  background-color: var(--color-primary);
  border-radius: var(--border-radius-button);
  padding: 0 16px;
  cursor: pointer;
  transition-timing-function: cubic-bezier(0, 0, 1, 1);
  transition-duration: 200ms;
  transition-property: background;
  white-space: nowrap;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;

  &:not([disabled]):hover {
    filter: brightness(87%);
  }

  &[disabled] {
    background-color: var(--color-disabled);
    color: rgb(var(--rgb-color-dark) / 50%) !important;
    cursor: default !important;
    pointer-events: none;

    &.outline {
      opacity: 0.8;
      border-color: var(--color-disabled);
      color: rgb(var(--rgb-color-tooltip) / 60%) !important;

      &:hover {
        background-color: rgb(var(--rgb-color-primary) / 8%) !important;
      }
    }
  }

  &.small {
    padding: 0 10px;
    height: 36px;
    font-size: 0.9rem;
  }

  &.medium {
    border: 3px solid red !important;
  }

  &.large {
    padding: 0 16px;
    height: 50px;
    font-size: 1.2rem;
  }

  &.xlarge {
    padding: 0 20px;
    height: 56px;
    font-size: 1.5rem;
  }

  &.secondary {
    background-color: var(--color-secondary);
    color: rgb(var(--rgb-background-color-primary));

    &:hover {
      background-color: rgb(var(--rgb-background-color-secondary-hover) / 80%);
      filter: brightness(130%);
    }
  }

  &.clear {
    background-color: transparent;
    color: var(--color-primary);

    &:hover {
      background-color: rgb(var(--rgb-background-color-clear-hover) / 8%) !important;
    }
  }

  &.outline {
    background-color: transparent !important;
    color: var(--color-primary);
    border: 1px solid var(--color-primary);

    &:hover {
      background-color: rgb(var(--rgb-background-color-clear-hover) / 8%);
    }
  }
}
