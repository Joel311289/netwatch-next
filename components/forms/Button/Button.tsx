import { MouseEvent, MouseEventHandler } from 'react';
import { ElementChildren, ElementHTML, ElementSkeleton } from '@/types';
import { withSkeleton } from '@/hoc/withSkeleton';
import { withTooltip } from '@/hoc/withTooltip';
import { classes } from '@/utils/helpers';
import styles from '@/components/forms/Button/Button.module.scss';

export type ButtonProps = Partial<typeof defaultProps> &
  ElementHTML &
  ElementChildren &
  ElementSkeleton & {
    size?: 'small' | 'medium' | 'large' | 'xlarge';
    ariaLabel?: string;
    tooltip?: string;
    secondary?: boolean;
    light?: boolean;
    dark?: boolean;
    clear?: boolean;
    outline?: boolean;
    disabled?: boolean;
    rounded?: boolean;
    onClick?: MouseEventHandler<HTMLButtonElement>;
  };

const defaultProps = {
  size: 'medium'
};

const Button = ({
  className,
  children,
  ariaLabel,
  size,
  secondary,
  clear,
  outline,
  disabled,
  tooltip,
  rounded,
  light,
  dark,
  onClick
}: ButtonProps) => {
  const handleClick = (event: MouseEvent<HTMLButtonElement>): void => {
    onClick && onClick(event);
  };

  return (
    <button
      aria-label={ariaLabel || tooltip}
      className={classes(
        styles.wrapper,
        size && styles[size],
        secondary && styles.secondary,
        light && styles.light,
        dark && styles.dark,
        clear && styles.clear,
        outline && styles.outline,
        rounded && styles.rounded,
        className
      )}
      disabled={disabled}
      onClick={handleClick}>
      {children}
    </button>
  );
};

Button.defaultProps = defaultProps;

const ButtonWithTooltip = withTooltip<ButtonProps>(Button);
const ButtonWithSkeleton = withSkeleton<ButtonProps>(ButtonWithTooltip, { variant: 'button' });

export default ButtonWithSkeleton;
