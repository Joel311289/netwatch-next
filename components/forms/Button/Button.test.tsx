import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Button from '@/components/forms/Button/Button';

describe('Tests Button component', () => {
  const textButton = 'Cancel';

  test('renders component correctly', () => {
    const { container } = render(<Button>Test</Button>);

    expect(container).toMatchSnapshot();
  });

  test('renders children correctly', () => {
    render(<Button>{textButton}</Button>);

    expect(screen.getByText(textButton)).toBeInTheDocument();
  });

  test('calls onClick prop correctly', async () => {
    const handleClick = jest.fn();

    render(<Button onClick={handleClick}>{textButton}</Button>);

    expect(handleClick).toHaveBeenCalledTimes(0);

    userEvent.click(screen.getByText(textButton));

    await waitFor(() => {
      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });

  test('renders disabled correctly', () => {
    render(<Button disabled>{textButton}</Button>);

    expect(screen.getByText(textButton)).toBeDisabled();
  });
});
