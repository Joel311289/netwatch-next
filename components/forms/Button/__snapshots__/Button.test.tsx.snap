// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Button component renders component correctly 1`] = `
<div>
  <button
    class="buttonWrapper medium"
  >
    Test
  </button>
</div>
`;
